README.md

```diff
@@ -0,0 +1,492 @@
+
+  - a function which takes any value of type `t a` and returns an array
+    containing a value of type `a` if possible (an empty array otherwise);
+    and
+
+  - the type of `a`.
+
+```haskell
+UnaryType :: String -> (* -> Boolean) -> (-> t _) -> (t a -> [a]) -> Type -> Type
+```
+
+For example:
+
+```javascript
+//    Maybe :: Type -> Type
+const Maybe = $.UnaryType(
+  'sanctuary-def/Maybe',
+  x => x != null && x['@@type'] === 'sanctuary-def/Maybe',
+  () => Nothing,
+  maybe => maybe.isJust ? [maybe.value] : []
+);
+
+//    Nothing :: Maybe a
+const Nothing = {
+  '@@type': 'sanctuary-def/Maybe',
+  'isJust': false,
+  'isNothing': true,
+  'toString': () => 'Nothing',
+};
+
+//    Just :: a -> Maybe a
+const Just = x => ({
+  '@@type': 'sanctuary-def/Maybe',
+  'isJust': true,
+  'isNothing': false,
+  'toString': () => 'Just(' + JSON.stringify(x) + ')',
+  'value': x,
+});
+
+//    fromMaybe :: a -> Maybe a -> a
+const fromMaybe = def('fromMaybe', {}, [a, Maybe(a)], (x, m) =>
+  m.isJust ? m.value : x
+);
+
+fromMaybe(0, Just(42));
+// => 42
+
+fromMaybe(0, Nothing);
+// => 0
+
+fromMaybe(0, Just('XXX'));
+// ! TypeError: ‘fromMaybe’ expected a value of type (Maybe Number) as its second argument; received Just("XXX")
+```
+
+#### `BinaryType`
+
+`BinaryType` is used to construct types with two type variables.
+
+To define a binary type `t a b` one must provide:
+
+  - the name of `t`;
+
+  - a predicate which accepts any JavaScript value and returns `true` if
+    (and only if) the value is a member of `t _ _`;
+
+  - a thunk which returns an arbitrary value of type `t _ _`;
+
+  - a function which takes any value of type `t a b` and returns an array
+    containing a value of type `a` if possible (an empty array otherwise);
+
+  - a function which takes any value of type `t a b` and returns an array
+    containing a value of type `b` if possible (an empty array otherwise);
+
+  - the type of `a`; and
+
+  - the type of `b`.
+
+```haskell
+BinaryType :: String -> (* -> Boolean) -> (-> t _ _) -> (t a b -> [a]) -> (t a b -> [b]) -> Type -> Type -> Type
+```
+
+For example:
+
+```javascript
+//    $Pair :: Type -> Type -> Type
+const $Pair = $.BinaryType(
+  'sanctuary-def/Pair',
+  x => x != null && x['@@type'] === 'sanctuary-def/Pair',
+  () => Pair(null, null),
+  pair => [pair[0]],
+  pair => [pair[1]]
+);
+
+//    Pair :: a -> b -> Pair a b
+const Pair = def('Pair', {}, [a, b], (x, y) => ({
+  '0': x,
+  '1': y,
+  '@@type': 'sanctuary-def/Pair',
+  'length': 2,
+  'toString': () => 'Pair(' + JSON.stringify(x) + ', ' + JSON.stringify(y) + ')',
+}));
+
+//    Rank :: Type
+const Rank = $.NullaryType(
+  'sanctuary-def/Rank',
+  x => typeof x === 'string' && /^([A23456789JQK]|10)$/.test(x),
+  'A'
+);
+
+//    Suit :: Type
+const Suit = $.NullaryType(
+  'sanctuary-def/Suit',
+  x => typeof x === 'string' && /^[\u2660\u2663\u2665\u2666]$/.test(x),
+  '\u2660'
+);
+
+//    Card :: Pair Rank Suit
+const Card = $Pair(Rank, Suit);
```

Comment:
In this case `$Pair` is the type constructor; `Pair` is the data constructor. Perhaps it would be clearer to use `Card :: Type` as the type signature. What do you think?